File: Algorithm/C++/930. Binary Subarrays With Sum.cpp
/*
 * Created on Wed Oct 31 2018 19:59:54
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */


// record prefix sum, applicable to not just 0 and 1
class Solution {
 public:
    int numSubarraysWithSum(vector<int>& A, int S) {
      int result = 0, curr_sum = 0;
      unordered_map<int, int> count;
      count[0] = 1;
      for(int num : A) {
        curr_sum += num;
        result += count[curr_sum - S];
        count[curr_sum] += 1;
      }
      return result;
    }
};

File: Algorithm/Python/930. Binary Subarrays With Sum.py
# -*- coding: utf-8 -*-
# Created on Tue Oct 30 2018 20:14:35
# Author: WuLC
# EMail: liangchaowu5@gmail.com

# record prefix sum，applicable to any number
from collections import defaultdict
class Solution(object):
    def numSubarraysWithSum(self, A, S):
        """
        :type A: List[int]
        :type S: int
        :rtype: int
        """
        count = defaultdict(int)
        count[0] = 1
        result, curr_sum = 0, 0
        for num in A:
            curr_sum += num
            result += count[curr_sum - S]
            count[curr_sum] += 1
        return result